src/tfr_mining/src/digging_queue.cpp

```diff
@@ -0,0 +1,120 @@
+#include "digging_queue.h"
+
+namespace tfr_mining
+{
+    // Must be a private node handle ("~")
+    DiggingQueue::DiggingQueue(ros::NodeHandle nh) : sets{}
+    {
+        // Make 4 digging sets, going from 45 degrees right to 90 degrees left
+        //  - The reasoning for this is so that we can always dump our excess
+        //    to the left. We'll fill in our previous holes a bit as we go, and
+        //    it ensures that we never put digging material in a place we'll dig
+        //    later
+        for (int j = 0; j < 4; j++)
+        {
+            DiggingSet set;
+            generateDigAndDump(nh, set, 3.14159265 / 4 - ((3.14159265 / 4) * j), 1);
+            generateDigAndDump(nh, set, 3.14159265 / 4 - ((3.14159265 / 4) * j), 2);
+            generateDigAndDump(nh, set, 3.14159265 / 4 - ((3.14159265 / 4) * j), 3);
+            generateDigAndDump(nh, set, 3.14159265 / 4 - ((3.14159265 / 4) * j), 4);
+            sets.push(set);
+        }
+    }
+
+    bool DiggingQueue::isEmpty()
+    {
+        return sets.empty();
+    }
+
+    DiggingSet DiggingQueue::popDiggingSet()
+    {
+        DiggingSet set = sets.front();
+        sets.pop();
+        return set;
+    }
+
+    void DiggingQueue::generateDigAndDump(ros::NodeHandle &nh, DiggingSet &set, double rotation, int dig_number)
+    {
+        // Ready (constant no matter what)
+        std::vector<double> ready_pos;
+        double ready_time;
+        if (!nh.getParam("positions/ready", ready_pos) || !nh.getParam("positions/ready_time", ready_time)) {
+            throw std::runtime_error("Error loading ready state");
```

Comment:
How are we dealing with exceptions? I don't remember talking a lot about error handling for things like this (mostly we've been using default parameters). Are you expecting this to crash the node?

Reply:
I have tried to avoid exceptions when I can. Usually I just print to ROS_WARN with configuration level stuff like this, it isn't a mission failing bug, it's just us being an idiot and mistyping something.

Reply:
I don't have a strong opinion here. This isn't something the robot can recover from without us going in and fixing something.

Reply:
Either way warns us that we are doing something stupid. Personally I would do the warning just to stay consistent with the rest of the code, but this version does the job so I don't think it's a big deal at all.

Reply:
I don't believe I used exceptions anywhere else, so I'll just print a warning instead and return early as to prevent unexpected behavior